study1/week2/Invert_Binary_Tree.java

```diff
@@ -0,0 +1,38 @@
+/**
+ * Definition for a binary tree node.
+ * public class TreeNode {
+ *     int val;
+ *     TreeNode left;
+ *     TreeNode right;
+ *     TreeNode() {}
+ *     TreeNode(int val) { this.val = val; }
+ *     TreeNode(int val, TreeNode left, TreeNode right) {
+ *         this.val = val;
+ *         this.left = left;
+ *         this.right = right;
+ *     }
+ * }
+ */
+class Invert_Binary_Tree {
+    public TreeNode invertTree(TreeNode root) {
+        if (root == null) {
+            return null;
+        }
+
+        TreeNode tmp = root.left;
+        root.left = root.right;
+        root.right = tmp;
+
+        invertTree(root.left);
+        invertTree(root.right);
+
+        return root;
+    }
+    /**
+     * Space Complexity : O(2^d), d is the depth of the TreeNode
```

Comment:
We usually talk about `d` in case it's a balanced tree. In the problem statement there seems no mention of this being balanced or not, so we should assume the worst case and assume it's unbalanced. So in these cases we should say the space complexity is `O(n)` - and talk in terms of number of nodes. And also, as this uses a recursive implementation, in the call stack there will be maximum n calls if this is unbalanced. Thus the space complexity will also be `O(n)` as well.